libtsuba/src/RDGPartHeader.h

```diff
@@ -16,11 +17,87 @@
 
 namespace tsuba {
 
-struct PropStorageInfo {
-  std::string name;
-  std::string path;
-  bool persist{false};
-  bool written_out{false};
+/// PropStorageInfo objects track the state of properties, and sanity check their
+/// transitions. N.b., It does not "DO" the transitions, this structure is purely
+/// for bookkeeping
+///
+/// Properties have 3 states:
+///  * Absent - exists in storage but is not in memory
+///  * Clean  - in memory and matches what is in storage
+///  * Dirty  - in memory but does not match what is in storage
+///
+/// The state machine looks like this:
+///
+/// EXISTING                                     NEW
+/// PROPERTY          modify                   PROPERTY
+///   |   +---------------------------------+     |
+///   |   |              +-------+  modify  |     |
+///   |   |     +------->| Clean +------+   |     |
+///   |   | load|        +-+-----+      |   |     |
+///   |   |     |          |   ^        v   v     |
+///   |  +------+-+        |   |       +-------+  |
+///   +->| Absent |<-------+   +-------+ Dirty |<-+
+///      +--------+  unload     write  +-------+
+///
+/// Properties either start out in storage as part of an RDG on disk
+/// (EXISTING PROPERTY) or start out in memory as part of an RDG in
+/// memory (NEW PROPERTY)
+class PropStorageInfo {
+  enum class State {
+    kAbsent,
+    kClean,
+    kDirty,
+  };
+
+public:
+  PropStorageInfo(std::string name)
+      : name_(std::move(name)), path_(), state_(State::kDirty) {}
+
+  PropStorageInfo(std::string name, std::string path)
+      : name_(std::move(name)),
+        path_(std::move(path)),
+        state_(State::kAbsent) {}
+
+  void NoteLoad() {
+    KATANA_LOG_ASSERT(state_ == State::kAbsent);
+    state_ = State::kClean;
+  }
+
+  void NoteModify() {
+    path_.clear();
+    state_ = State::kDirty;
+  }
+
+  void NoteWrite(std::string_view new_path) {
+    KATANA_LOG_ASSERT(state_ == State::kDirty);
+    path_ = new_path;
+    state_ = State::kClean;
+  }
+
+  void NoteUnload() {
+    KATANA_LOG_ASSERT(state_ == State::kClean);
+    state_ = State::kAbsent;
+  }
```

Comment:
These functions make a lot of sense, but I'm not crazy about the `Note*` naming convention.  Can we name them more for the state transitions they represent?